##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# Text classification with TensorFlow Hub: Movie reviews

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
</table>

This notebook classifies movie reviews as *positive* or *negative* using the text of the review. This is an example of *binary*—or two-class—classification, an important and widely applicable kind of machine learning problem.

The tutorial demonstrates the basic application of transfer learning with TensorFlow Hub and Keras.

We'll use the [IMDB dataset](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb) that contains the text of 50,000 movie reviews from the [Internet Movie Database](https://www.imdb.com/). These are split into 25,000 reviews for training and 25,000 reviews for testing. The training and testing sets are *balanced*, meaning they contain an equal number of positive and negative reviews. 

This notebook uses [tf.keras](https://www.tensorflow.org/guide/keras), a high-level API to build and train models in TensorFlow, and [TensorFlow Hub](https://www.tensorflow.org/hub), a library and platform for transfer learning. For a more advanced text classification tutorial using `tf.keras`, see the [MLCC Text Classification Guide](https://developers.google.com/machine-learning/guides/text-classification/).

In [3]:
import numpy as np

import tensorflow as tf

!pip install -q tensorflow-hub
!pip install -q tensorflow-datasets
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.experimental.list_physical_devices("GPU") else "NOT AVAILABLE")

Version:  2.3.0
Eager mode:  True
Hub version:  0.8.0
GPU is available


## Download the IMDB dataset

The IMDB dataset is available on [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) or on [TensorFlow datasets](https://www.tensorflow.org/datasets). The following code downloads the IMDB dataset to your machine (or the colab runtime):

In [4]:
# Split the training set into 60% and 40%, so we'll end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incomplete452INS/imdb_reviews-train.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incomplete452INS/imdb_reviews-test.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incomplete452INS/imdb_reviews-unsupervised.tfrecord


Dataset imdb_reviews downloaded and prepared to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0. Subsequent calls will reuse this data.


## Explore the data 

Let's take a moment to understand the format of the data. Each example is a sentence representing the movie review and a corresponding label. The sentence is not preprocessed in any way. The label is an integer value of either 0 or 1, where 0 is a negative review, and 1 is a positive review.

Let's print first 10 examples.

In [5]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

Let's also print the first 10 labels.

In [6]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## Build the model

The neural network is created by stacking layers—this requires three main architectural decisions:

* How to represent the text?
* How many layers to use in the model?
* How many *hidden units* to use for each layer?

In this example, the input data consists of sentences. The labels to predict are either 0 or 1.

One way to represent the text is to convert sentences into embeddings vectors. We can use a pre-trained text embedding as the first layer, which will have three advantages:

*   we don't have to worry about text preprocessing,
*   we can benefit from transfer learning,
*   the embedding has a fixed size, so it's simpler to process.

For this example we will use a **pre-trained text embedding model** from [TensorFlow Hub](https://www.tensorflow.org/hub) called [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1).

There are three other pre-trained models to test for the sake of this tutorial:

* [google/tf2-preview/gnews-swivel-20dim-with-oov/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim-with-oov/1) - same as [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1), but with 2.5% vocabulary converted to OOV buckets. This can help if vocabulary of the task and vocabulary of the model don't fully overlap.
* [google/tf2-preview/nnlm-en-dim50/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim50/1) - A much larger model with ~1M vocabulary size and 50 dimensions.
* [google/tf2-preview/nnlm-en-dim128/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim128/1) - Even larger model with ~1M vocabulary size and 128 dimensions.

Let's first create a Keras layer that uses a TensorFlow Hub model to embed the sentences, and try it out on a couple of input examples. Note that no matter the length of the input text, the output shape of the embeddings is: `(num_examples, embedding_dimension)`.

In [7]:
embedding = "https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 20), dtype=float32, numpy=
array([[ 1.765786  , -3.882232  ,  3.9134233 , -1.5557289 , -3.3362343 ,
        -1.7357955 , -1.9954445 ,  1.2989551 ,  5.081598  , -1.1041286 ,
        -2.0503852 , -0.72675157, -0.65675956,  0.24436149, -3.7208383 ,
         2.0954835 ,  2.2969332 , -2.0689783 , -2.9489717 , -1.1315987 ],
       [ 1.8804485 , -2.5852382 ,  3.4066997 ,  1.0982676 , -4.056685  ,
        -4.891284  , -2.785554  ,  1.3874227 ,  3.8476458 , -0.9256538 ,
        -1.896706  ,  1.2113281 ,  0.11474707,  0.76209456, -4.8791065 ,
         2.906149  ,  4.7087674 , -2.3652055 , -3.5015898 , -1.6390051 ],
       [ 0.71152234, -0.6353217 ,  1.7385626 , -1.1168286 , -0.5451594 ,
        -1.1808156 ,  0.09504455,  1.4653089 ,  0.66059524,  0.79308075,
        -2.2268345 ,  0.07446612, -1.4075904 , -0.70645386, -1.907037  ,
         1.4419787 ,  1.9551861 , -0.42660055, -2.8022065 ,  0.43727064]],
      dtype=float32)>

Let's now build the full model:

In [8]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 20)                400020    
_________________________________________________________________
dense (Dense)                (None, 16)                336       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 400,373
Trainable params: 400,373
Non-trainable params: 0
_________________________________________________________________


The layers are stacked sequentially to build the classifier:

1. The first layer is a TensorFlow Hub layer. This layer uses a pre-trained Saved Model to map a sentence into its embedding vector. The pre-trained text embedding model that we are using ([google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1)) splits the sentence into tokens, embeds each token and then combines the embedding. The resulting dimensions are: `(num_examples, embedding_dimension)`.
2. This fixed-length output vector is piped through a fully-connected (`Dense`) layer with 16 hidden units.
3. The last layer is densely connected with a single output node.

Let's compile the model.

### Loss function and optimizer

A model needs a loss function and an optimizer for training. Since this is a binary classification problem and the model outputs logits (a single-unit layer with a linear activation), we'll use the `binary_crossentropy` loss function.

This isn't the only choice for a loss function, you could, for instance, choose `mean_squared_error`. But, generally, `binary_crossentropy` is better for dealing with probabilities—it measures the "distance" between probability distributions, or in our case, between the ground-truth distribution and the predictions.

Later, when we are exploring regression problems (say, to predict the price of a house), we will see how to use another loss function called mean squared error.

Now, configure the model to use an optimizer and a loss function:

In [9]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## Train the model

Train the model for 20 epochs in mini-batches of 512 samples. This is 20 iterations over all samples in the `x_train` and `y_train` tensors. While training, monitor the model's loss and accuracy on the 10,000 samples from the validation set:

In [10]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=20,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/20


 1/30 [>.............................] - ETA: 0s - loss: 1.1224 - accuracy: 0.3965

 3/30 [==>...........................] - ETA: 0s - loss: 1.0944 - accuracy: 0.4121

 5/30 [====>.........................] - ETA: 0s - loss: 1.0744 - accuracy: 0.4191

 7/30 [======>.......................] - ETA: 0s - loss: 1.0620 - accuracy: 0.4322

 9/30 [========>.....................] - ETA: 0s - loss: 1.0518 - accuracy: 0.4373

11/30 [==========>...................] - ETA: 0s - loss: 1.0398 - accuracy: 0.4409

13/30 [============>.................] - ETA: 0s - loss: 1.0279 - accuracy: 0.4422

15/30 [==============>...............] - ETA: 0s - loss: 1.0156 - accuracy: 0.4456

17/30 [================>.............] - ETA: 0s - loss: 1.0018 - accuracy: 0.4484

19/30 [==================>...........] - ETA: 0s - loss: 0.9899 - accuracy: 0.4508

21/30 [====================>.........] - ETA: 0s - loss: 0.9808 - accuracy: 0.4519

23/30 [======================>.......] - ETA: 0s - loss: 0.9658 - accuracy: 0.4565

25/30 [========================>.....] - ETA: 0s - loss: 0.9535 - accuracy: 0.4603

27/30 [==========================>...] - ETA: 0s - loss: 0.9422 - accuracy: 0.4630

29/30 [============================>.] - ETA: 0s - loss: 0.9316 - accuracy: 0.4657

30/30 [==============================] - 2s 54ms/step - loss: 0.9307 - accuracy: 0.4662 - val_loss: 0.7834 - val_accuracy: 0.5065


Epoch 2/20


 1/30 [>.............................] - ETA: 0s - loss: 0.7646 - accuracy: 0.5156

 3/30 [==>...........................] - ETA: 0s - loss: 0.7666 - accuracy: 0.5221

 5/30 [====>.........................] - ETA: 0s - loss: 0.7656 - accuracy: 0.5211

 7/30 [======>.......................] - ETA: 0s - loss: 0.7638 - accuracy: 0.5193

 9/30 [========>.....................] - ETA: 0s - loss: 0.7594 - accuracy: 0.5200

11/30 [==========>...................] - ETA: 0s - loss: 0.7560 - accuracy: 0.5206

13/30 [============>.................] - ETA: 0s - loss: 0.7509 - accuracy: 0.5266

15/30 [==============>...............] - ETA: 0s - loss: 0.7474 - accuracy: 0.5277

17/30 [================>.............] - ETA: 0s - loss: 0.7445 - accuracy: 0.5271

19/30 [==================>...........] - ETA: 0s - loss: 0.7419 - accuracy: 0.5270

21/30 [====================>.........] - ETA: 0s - loss: 0.7375 - accuracy: 0.5299

23/30 [======================>.......] - ETA: 0s - loss: 0.7344 - accuracy: 0.5319

25/30 [========================>.....] - ETA: 0s - loss: 0.7321 - accuracy: 0.5337

27/30 [==========================>...] - ETA: 0s - loss: 0.7304 - accuracy: 0.5347

29/30 [============================>.] - ETA: 0s - loss: 0.7278 - accuracy: 0.5355

30/30 [==============================] - 1s 49ms/step - loss: 0.7267 - accuracy: 0.5361 - val_loss: 0.6816 - val_accuracy: 0.5643


Epoch 3/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6733 - accuracy: 0.5586

 3/30 [==>...........................] - ETA: 0s - loss: 0.6715 - accuracy: 0.5547

 5/30 [====>.........................] - ETA: 0s - loss: 0.6730 - accuracy: 0.5594

 7/30 [======>.......................] - ETA: 0s - loss: 0.6757 - accuracy: 0.5664

 9/30 [========>.....................] - ETA: 0s - loss: 0.6723 - accuracy: 0.5651

11/30 [==========>...................] - ETA: 0s - loss: 0.6695 - accuracy: 0.5685

13/30 [============>.................] - ETA: 0s - loss: 0.6669 - accuracy: 0.5705

15/30 [==============>...............] - ETA: 0s - loss: 0.6656 - accuracy: 0.5698

17/30 [================>.............] - ETA: 0s - loss: 0.6634 - accuracy: 0.5756

19/30 [==================>...........] - ETA: 0s - loss: 0.6615 - accuracy: 0.5757

21/30 [====================>.........] - ETA: 0s - loss: 0.6601 - accuracy: 0.5784

23/30 [======================>.......] - ETA: 0s - loss: 0.6569 - accuracy: 0.5808

25/30 [========================>.....] - ETA: 0s - loss: 0.6566 - accuracy: 0.5784

27/30 [==========================>...] - ETA: 0s - loss: 0.6551 - accuracy: 0.5803

29/30 [============================>.] - ETA: 0s - loss: 0.6527 - accuracy: 0.5845

30/30 [==============================] - 1s 49ms/step - loss: 0.6527 - accuracy: 0.5848 - val_loss: 0.6294 - val_accuracy: 0.6098


Epoch 4/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6134 - accuracy: 0.6309

 3/30 [==>...........................] - ETA: 0s - loss: 0.6234 - accuracy: 0.6113

 5/30 [====>.........................] - ETA: 0s - loss: 0.6172 - accuracy: 0.6141

 7/30 [======>.......................] - ETA: 0s - loss: 0.6152 - accuracy: 0.6175

 9/30 [========>.....................] - ETA: 0s - loss: 0.6144 - accuracy: 0.6241

11/30 [==========>...................] - ETA: 0s - loss: 0.6134 - accuracy: 0.6271

13/30 [============>.................] - ETA: 0s - loss: 0.6104 - accuracy: 0.6295

15/30 [==============>...............] - ETA: 0s - loss: 0.6096 - accuracy: 0.6294

17/30 [================>.............] - ETA: 0s - loss: 0.6095 - accuracy: 0.6271

19/30 [==================>...........] - ETA: 0s - loss: 0.6096 - accuracy: 0.6273

21/30 [====================>.........] - ETA: 0s - loss: 0.6083 - accuracy: 0.6293

23/30 [======================>.......] - ETA: 0s - loss: 0.6071 - accuracy: 0.6296

25/30 [========================>.....] - ETA: 0s - loss: 0.6059 - accuracy: 0.6298

27/30 [==========================>...] - ETA: 0s - loss: 0.6039 - accuracy: 0.6332

29/30 [============================>.] - ETA: 0s - loss: 0.6022 - accuracy: 0.6367

30/30 [==============================] - 1s 49ms/step - loss: 0.6021 - accuracy: 0.6364 - val_loss: 0.5885 - val_accuracy: 0.6691


Epoch 5/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5869 - accuracy: 0.6445

 3/30 [==>...........................] - ETA: 0s - loss: 0.5797 - accuracy: 0.6758

 5/30 [====>.........................] - ETA: 0s - loss: 0.5782 - accuracy: 0.6785

 7/30 [======>.......................] - ETA: 0s - loss: 0.5704 - accuracy: 0.6814

 9/30 [========>.....................] - ETA: 0s - loss: 0.5685 - accuracy: 0.6766

11/30 [==========>...................] - ETA: 0s - loss: 0.5695 - accuracy: 0.6701

13/30 [============>.................] - ETA: 0s - loss: 0.5697 - accuracy: 0.6677

15/30 [==============>...............] - ETA: 0s - loss: 0.5690 - accuracy: 0.6664

17/30 [================>.............] - ETA: 0s - loss: 0.5675 - accuracy: 0.6714

19/30 [==================>...........] - ETA: 0s - loss: 0.5654 - accuracy: 0.6767

21/30 [====================>.........] - ETA: 0s - loss: 0.5649 - accuracy: 0.6802

23/30 [======================>.......] - ETA: 0s - loss: 0.5632 - accuracy: 0.6821

25/30 [========================>.....] - ETA: 0s - loss: 0.5615 - accuracy: 0.6823

27/30 [==========================>...] - ETA: 0s - loss: 0.5613 - accuracy: 0.6827

29/30 [============================>.] - ETA: 0s - loss: 0.5596 - accuracy: 0.6838

30/30 [==============================] - 1s 48ms/step - loss: 0.5594 - accuracy: 0.6837 - val_loss: 0.5497 - val_accuracy: 0.6937


Epoch 6/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5437 - accuracy: 0.7031

 3/30 [==>...........................] - ETA: 0s - loss: 0.5272 - accuracy: 0.7253

 5/30 [====>.........................] - ETA: 0s - loss: 0.5352 - accuracy: 0.7164

 7/30 [======>.......................] - ETA: 0s - loss: 0.5316 - accuracy: 0.7160

 9/30 [========>.....................] - ETA: 0s - loss: 0.5306 - accuracy: 0.7190

11/30 [==========>...................] - ETA: 0s - loss: 0.5286 - accuracy: 0.7177

13/30 [============>.................] - ETA: 0s - loss: 0.5276 - accuracy: 0.7227

15/30 [==============>...............] - ETA: 0s - loss: 0.5264 - accuracy: 0.7242

17/30 [================>.............] - ETA: 0s - loss: 0.5246 - accuracy: 0.7256

19/30 [==================>...........] - ETA: 0s - loss: 0.5218 - accuracy: 0.7282

21/30 [====================>.........] - ETA: 0s - loss: 0.5227 - accuracy: 0.7265

23/30 [======================>.......] - ETA: 0s - loss: 0.5229 - accuracy: 0.7240

25/30 [========================>.....] - ETA: 0s - loss: 0.5208 - accuracy: 0.7260

27/30 [==========================>...] - ETA: 0s - loss: 0.5203 - accuracy: 0.7258

29/30 [============================>.] - ETA: 0s - loss: 0.5200 - accuracy: 0.7265

30/30 [==============================] - 1s 48ms/step - loss: 0.5195 - accuracy: 0.7267 - val_loss: 0.5182 - val_accuracy: 0.7501


Epoch 7/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5332 - accuracy: 0.7559

 3/30 [==>...........................] - ETA: 0s - loss: 0.5230 - accuracy: 0.7383

 5/30 [====>.........................] - ETA: 0s - loss: 0.5159 - accuracy: 0.7402

 7/30 [======>.......................] - ETA: 0s - loss: 0.5051 - accuracy: 0.7506

 9/30 [========>.....................] - ETA: 0s - loss: 0.4996 - accuracy: 0.7485

11/30 [==========>...................] - ETA: 0s - loss: 0.4959 - accuracy: 0.7475

13/30 [============>.................] - ETA: 0s - loss: 0.4940 - accuracy: 0.7486

15/30 [==============>...............] - ETA: 0s - loss: 0.4932 - accuracy: 0.7477

17/30 [================>.............] - ETA: 0s - loss: 0.4910 - accuracy: 0.7494

19/30 [==================>...........] - ETA: 0s - loss: 0.4913 - accuracy: 0.7507

21/30 [====================>.........] - ETA: 0s - loss: 0.4892 - accuracy: 0.7533

23/30 [======================>.......] - ETA: 0s - loss: 0.4870 - accuracy: 0.7555

25/30 [========================>.....] - ETA: 0s - loss: 0.4865 - accuracy: 0.7563

27/30 [==========================>...] - ETA: 0s - loss: 0.4835 - accuracy: 0.7582

29/30 [============================>.] - ETA: 0s - loss: 0.4822 - accuracy: 0.7592

30/30 [==============================] - 1s 49ms/step - loss: 0.4817 - accuracy: 0.7598 - val_loss: 0.4821 - val_accuracy: 0.7482


Epoch 8/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4795 - accuracy: 0.7578

 3/30 [==>...........................] - ETA: 0s - loss: 0.4512 - accuracy: 0.7767

 5/30 [====>.........................] - ETA: 0s - loss: 0.4550 - accuracy: 0.7766

 7/30 [======>.......................] - ETA: 0s - loss: 0.4539 - accuracy: 0.7768

 9/30 [========>.....................] - ETA: 0s - loss: 0.4534 - accuracy: 0.7836

11/30 [==========>...................] - ETA: 0s - loss: 0.4509 - accuracy: 0.7905

13/30 [============>.................] - ETA: 0s - loss: 0.4501 - accuracy: 0.7894

15/30 [==============>...............] - ETA: 0s - loss: 0.4505 - accuracy: 0.7871

17/30 [================>.............] - ETA: 0s - loss: 0.4515 - accuracy: 0.7824

19/30 [==================>...........] - ETA: 0s - loss: 0.4514 - accuracy: 0.7826

21/30 [====================>.........] - ETA: 0s - loss: 0.4520 - accuracy: 0.7813

23/30 [======================>.......] - ETA: 0s - loss: 0.4493 - accuracy: 0.7828

25/30 [========================>.....] - ETA: 0s - loss: 0.4481 - accuracy: 0.7849

27/30 [==========================>...] - ETA: 0s - loss: 0.4469 - accuracy: 0.7863

29/30 [============================>.] - ETA: 0s - loss: 0.4452 - accuracy: 0.7862

30/30 [==============================] - 1s 48ms/step - loss: 0.4449 - accuracy: 0.7866 - val_loss: 0.4503 - val_accuracy: 0.7707


Epoch 9/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4178 - accuracy: 0.7793

 3/30 [==>...........................] - ETA: 0s - loss: 0.4167 - accuracy: 0.7923

 5/30 [====>.........................] - ETA: 0s - loss: 0.4231 - accuracy: 0.7953

 7/30 [======>.......................] - ETA: 0s - loss: 0.4174 - accuracy: 0.8039

 9/30 [========>.....................] - ETA: 0s - loss: 0.4164 - accuracy: 0.8088

11/30 [==========>...................] - ETA: 0s - loss: 0.4165 - accuracy: 0.8077

13/30 [============>.................] - ETA: 0s - loss: 0.4162 - accuracy: 0.8119

15/30 [==============>...............] - ETA: 0s - loss: 0.4156 - accuracy: 0.8120

17/30 [================>.............] - ETA: 0s - loss: 0.4154 - accuracy: 0.8096

19/30 [==================>...........] - ETA: 0s - loss: 0.4139 - accuracy: 0.8103

21/30 [====================>.........] - ETA: 0s - loss: 0.4132 - accuracy: 0.8118

23/30 [======================>.......] - ETA: 0s - loss: 0.4120 - accuracy: 0.8133

25/30 [========================>.....] - ETA: 0s - loss: 0.4095 - accuracy: 0.8155

27/30 [==========================>...] - ETA: 0s - loss: 0.4096 - accuracy: 0.8148

29/30 [============================>.] - ETA: 0s - loss: 0.4083 - accuracy: 0.8142

30/30 [==============================] - 1s 49ms/step - loss: 0.4077 - accuracy: 0.8141 - val_loss: 0.4209 - val_accuracy: 0.7933


Epoch 10/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3784 - accuracy: 0.8320

 3/30 [==>...........................] - ETA: 0s - loss: 0.3794 - accuracy: 0.8171

 5/30 [====>.........................] - ETA: 0s - loss: 0.3794 - accuracy: 0.8254

 7/30 [======>.......................] - ETA: 0s - loss: 0.3752 - accuracy: 0.8329

 9/30 [========>.....................] - ETA: 0s - loss: 0.3771 - accuracy: 0.8336

11/30 [==========>...................] - ETA: 0s - loss: 0.3786 - accuracy: 0.8335

13/30 [============>.................] - ETA: 0s - loss: 0.3784 - accuracy: 0.8358

15/30 [==============>...............] - ETA: 0s - loss: 0.3792 - accuracy: 0.8310

17/30 [================>.............] - ETA: 0s - loss: 0.3785 - accuracy: 0.8295

19/30 [==================>...........] - ETA: 0s - loss: 0.3777 - accuracy: 0.8307

21/30 [====================>.........] - ETA: 0s - loss: 0.3790 - accuracy: 0.8316

23/30 [======================>.......] - ETA: 0s - loss: 0.3770 - accuracy: 0.8325

25/30 [========================>.....] - ETA: 0s - loss: 0.3751 - accuracy: 0.8327

27/30 [==========================>...] - ETA: 0s - loss: 0.3752 - accuracy: 0.8332

29/30 [============================>.] - ETA: 0s - loss: 0.3746 - accuracy: 0.8334

30/30 [==============================] - 1s 48ms/step - loss: 0.3744 - accuracy: 0.8335 - val_loss: 0.3944 - val_accuracy: 0.8122


Epoch 11/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3653 - accuracy: 0.8398

 3/30 [==>...........................] - ETA: 0s - loss: 0.3632 - accuracy: 0.8457

 5/30 [====>.........................] - ETA: 0s - loss: 0.3554 - accuracy: 0.8457

 7/30 [======>.......................] - ETA: 0s - loss: 0.3546 - accuracy: 0.8485

 9/30 [========>.....................] - ETA: 0s - loss: 0.3522 - accuracy: 0.8516

11/30 [==========>...................] - ETA: 0s - loss: 0.3521 - accuracy: 0.8530

13/30 [============>.................] - ETA: 0s - loss: 0.3493 - accuracy: 0.8534

15/30 [==============>...............] - ETA: 0s - loss: 0.3493 - accuracy: 0.8490

17/30 [================>.............] - ETA: 0s - loss: 0.3493 - accuracy: 0.8470

19/30 [==================>...........] - ETA: 0s - loss: 0.3483 - accuracy: 0.8488

21/30 [====================>.........] - ETA: 0s - loss: 0.3480 - accuracy: 0.8498

23/30 [======================>.......] - ETA: 0s - loss: 0.3469 - accuracy: 0.8505

25/30 [========================>.....] - ETA: 0s - loss: 0.3461 - accuracy: 0.8507

27/30 [==========================>...] - ETA: 0s - loss: 0.3445 - accuracy: 0.8517

29/30 [============================>.] - ETA: 0s - loss: 0.3435 - accuracy: 0.8506

30/30 [==============================] - 1s 48ms/step - loss: 0.3436 - accuracy: 0.8501 - val_loss: 0.3734 - val_accuracy: 0.8327


Epoch 12/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3498 - accuracy: 0.8594

 3/30 [==>...........................] - ETA: 0s - loss: 0.3217 - accuracy: 0.8678

 5/30 [====>.........................] - ETA: 0s - loss: 0.3203 - accuracy: 0.8687

 7/30 [======>.......................] - ETA: 0s - loss: 0.3169 - accuracy: 0.8697

 9/30 [========>.....................] - ETA: 0s - loss: 0.3211 - accuracy: 0.8665

11/30 [==========>...................] - ETA: 0s - loss: 0.3230 - accuracy: 0.8642

13/30 [============>.................] - ETA: 0s - loss: 0.3212 - accuracy: 0.8655

15/30 [==============>...............] - ETA: 0s - loss: 0.3204 - accuracy: 0.8641

17/30 [================>.............] - ETA: 0s - loss: 0.3192 - accuracy: 0.8621

19/30 [==================>...........] - ETA: 0s - loss: 0.3189 - accuracy: 0.8629

21/30 [====================>.........] - ETA: 0s - loss: 0.3199 - accuracy: 0.8629

23/30 [======================>.......] - ETA: 0s - loss: 0.3197 - accuracy: 0.8623

25/30 [========================>.....] - ETA: 0s - loss: 0.3180 - accuracy: 0.8637

27/30 [==========================>...] - ETA: 0s - loss: 0.3180 - accuracy: 0.8639

29/30 [============================>.] - ETA: 0s - loss: 0.3166 - accuracy: 0.8648

30/30 [==============================] - 1s 49ms/step - loss: 0.3164 - accuracy: 0.8651 - val_loss: 0.3559 - val_accuracy: 0.8434


Epoch 13/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3106 - accuracy: 0.8613

 3/30 [==>...........................] - ETA: 0s - loss: 0.3072 - accuracy: 0.8646

 5/30 [====>.........................] - ETA: 0s - loss: 0.3084 - accuracy: 0.8680

 7/30 [======>.......................] - ETA: 0s - loss: 0.3049 - accuracy: 0.8694

 9/30 [========>.....................] - ETA: 0s - loss: 0.2999 - accuracy: 0.8713

11/30 [==========>...................] - ETA: 0s - loss: 0.2962 - accuracy: 0.8736

13/30 [============>.................] - ETA: 0s - loss: 0.2965 - accuracy: 0.8741

15/30 [==============>...............] - ETA: 0s - loss: 0.2954 - accuracy: 0.8742

17/30 [================>.............] - ETA: 0s - loss: 0.2937 - accuracy: 0.8742

19/30 [==================>...........] - ETA: 0s - loss: 0.2935 - accuracy: 0.8743

21/30 [====================>.........] - ETA: 0s - loss: 0.2933 - accuracy: 0.8744

23/30 [======================>.......] - ETA: 0s - loss: 0.2919 - accuracy: 0.8752

25/30 [========================>.....] - ETA: 0s - loss: 0.2916 - accuracy: 0.8760

27/30 [==========================>...] - ETA: 0s - loss: 0.2918 - accuracy: 0.8765

29/30 [============================>.] - ETA: 0s - loss: 0.2915 - accuracy: 0.8766

30/30 [==============================] - 1s 48ms/step - loss: 0.2919 - accuracy: 0.8768 - val_loss: 0.3413 - val_accuracy: 0.8492


Epoch 14/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2507 - accuracy: 0.9004

 3/30 [==>...........................] - ETA: 0s - loss: 0.2721 - accuracy: 0.8945

 5/30 [====>.........................] - ETA: 0s - loss: 0.2755 - accuracy: 0.8945

 7/30 [======>.......................] - ETA: 0s - loss: 0.2722 - accuracy: 0.8956

 9/30 [========>.....................] - ETA: 0s - loss: 0.2758 - accuracy: 0.8900

11/30 [==========>...................] - ETA: 0s - loss: 0.2766 - accuracy: 0.8874

13/30 [============>.................] - ETA: 0s - loss: 0.2779 - accuracy: 0.8849

15/30 [==============>...............] - ETA: 0s - loss: 0.2759 - accuracy: 0.8870

17/30 [================>.............] - ETA: 0s - loss: 0.2746 - accuracy: 0.8887

19/30 [==================>...........] - ETA: 0s - loss: 0.2730 - accuracy: 0.8892

21/30 [====================>.........] - ETA: 0s - loss: 0.2718 - accuracy: 0.8896

23/30 [======================>.......] - ETA: 0s - loss: 0.2712 - accuracy: 0.8895

25/30 [========================>.....] - ETA: 0s - loss: 0.2706 - accuracy: 0.8899

27/30 [==========================>...] - ETA: 0s - loss: 0.2719 - accuracy: 0.8894

29/30 [============================>.] - ETA: 0s - loss: 0.2691 - accuracy: 0.8906

30/30 [==============================] - 1s 48ms/step - loss: 0.2701 - accuracy: 0.8899 - val_loss: 0.3295 - val_accuracy: 0.8563


Epoch 15/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2503 - accuracy: 0.9043

 3/30 [==>...........................] - ETA: 0s - loss: 0.2533 - accuracy: 0.8978

 5/30 [====>.........................] - ETA: 0s - loss: 0.2515 - accuracy: 0.8969

 7/30 [======>.......................] - ETA: 0s - loss: 0.2509 - accuracy: 0.8982

 9/30 [========>.....................] - ETA: 0s - loss: 0.2541 - accuracy: 0.8971

11/30 [==========>...................] - ETA: 0s - loss: 0.2537 - accuracy: 0.8991

13/30 [============>.................] - ETA: 0s - loss: 0.2545 - accuracy: 0.8983

15/30 [==============>...............] - ETA: 0s - loss: 0.2518 - accuracy: 0.8991

17/30 [================>.............] - ETA: 0s - loss: 0.2505 - accuracy: 0.8994

19/30 [==================>...........] - ETA: 0s - loss: 0.2499 - accuracy: 0.9002

21/30 [====================>.........] - ETA: 0s - loss: 0.2515 - accuracy: 0.8993

23/30 [======================>.......] - ETA: 0s - loss: 0.2501 - accuracy: 0.9003

25/30 [========================>.....] - ETA: 0s - loss: 0.2488 - accuracy: 0.9015

27/30 [==========================>...] - ETA: 0s - loss: 0.2479 - accuracy: 0.9020

29/30 [============================>.] - ETA: 0s - loss: 0.2505 - accuracy: 0.8995

30/30 [==============================] - 1s 48ms/step - loss: 0.2507 - accuracy: 0.8993 - val_loss: 0.3216 - val_accuracy: 0.8629


Epoch 16/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2364 - accuracy: 0.9121

 3/30 [==>...........................] - ETA: 0s - loss: 0.2276 - accuracy: 0.9069

 5/30 [====>.........................] - ETA: 0s - loss: 0.2345 - accuracy: 0.9090

 7/30 [======>.......................] - ETA: 0s - loss: 0.2329 - accuracy: 0.9068

 9/30 [========>.....................] - ETA: 0s - loss: 0.2352 - accuracy: 0.9052

11/30 [==========>...................] - ETA: 0s - loss: 0.2352 - accuracy: 0.9045

13/30 [============>.................] - ETA: 0s - loss: 0.2368 - accuracy: 0.9049

15/30 [==============>...............] - ETA: 0s - loss: 0.2376 - accuracy: 0.9051

17/30 [================>.............] - ETA: 0s - loss: 0.2383 - accuracy: 0.9051

19/30 [==================>...........] - ETA: 0s - loss: 0.2368 - accuracy: 0.9064

21/30 [====================>.........] - ETA: 0s - loss: 0.2354 - accuracy: 0.9063

23/30 [======================>.......] - ETA: 0s - loss: 0.2358 - accuracy: 0.9065

25/30 [========================>.....] - ETA: 0s - loss: 0.2355 - accuracy: 0.9063

27/30 [==========================>...] - ETA: 0s - loss: 0.2336 - accuracy: 0.9074

29/30 [============================>.] - ETA: 0s - loss: 0.2332 - accuracy: 0.9075

30/30 [==============================] - 1s 48ms/step - loss: 0.2327 - accuracy: 0.9077 - val_loss: 0.3145 - val_accuracy: 0.8655


Epoch 17/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2152 - accuracy: 0.9336

 3/30 [==>...........................] - ETA: 0s - loss: 0.2148 - accuracy: 0.9251

 5/30 [====>.........................] - ETA: 0s - loss: 0.2148 - accuracy: 0.9160

 7/30 [======>.......................] - ETA: 0s - loss: 0.2224 - accuracy: 0.9110

 9/30 [========>.....................] - ETA: 0s - loss: 0.2197 - accuracy: 0.9147

11/30 [==========>...................] - ETA: 0s - loss: 0.2189 - accuracy: 0.9169

13/30 [============>.................] - ETA: 0s - loss: 0.2211 - accuracy: 0.9150

15/30 [==============>...............] - ETA: 0s - loss: 0.2201 - accuracy: 0.9147

17/30 [================>.............] - ETA: 0s - loss: 0.2219 - accuracy: 0.9131

19/30 [==================>...........] - ETA: 0s - loss: 0.2222 - accuracy: 0.9139

21/30 [====================>.........] - ETA: 0s - loss: 0.2212 - accuracy: 0.9142

23/30 [======================>.......] - ETA: 0s - loss: 0.2186 - accuracy: 0.9161

25/30 [========================>.....] - ETA: 0s - loss: 0.2175 - accuracy: 0.9166

27/30 [==========================>...] - ETA: 0s - loss: 0.2181 - accuracy: 0.9162

29/30 [============================>.] - ETA: 0s - loss: 0.2169 - accuracy: 0.9168

30/30 [==============================] - 1s 49ms/step - loss: 0.2170 - accuracy: 0.9163 - val_loss: 0.3079 - val_accuracy: 0.8606


Epoch 18/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2210 - accuracy: 0.9121

 3/30 [==>...........................] - ETA: 0s - loss: 0.2023 - accuracy: 0.9277

 5/30 [====>.........................] - ETA: 0s - loss: 0.2020 - accuracy: 0.9285

 7/30 [======>.......................] - ETA: 0s - loss: 0.1966 - accuracy: 0.9277

 9/30 [========>.....................] - ETA: 0s - loss: 0.1951 - accuracy: 0.9282

11/30 [==========>...................] - ETA: 0s - loss: 0.1943 - accuracy: 0.9268

13/30 [============>.................] - ETA: 0s - loss: 0.1971 - accuracy: 0.9244

15/30 [==============>...............] - ETA: 0s - loss: 0.1991 - accuracy: 0.9230

17/30 [================>.............] - ETA: 0s - loss: 0.1990 - accuracy: 0.9236

19/30 [==================>...........] - ETA: 0s - loss: 0.2004 - accuracy: 0.9238

21/30 [====================>.........] - ETA: 0s - loss: 0.2019 - accuracy: 0.9240

23/30 [======================>.......] - ETA: 0s - loss: 0.2029 - accuracy: 0.9234

25/30 [========================>.....] - ETA: 0s - loss: 0.2018 - accuracy: 0.9238

27/30 [==========================>...] - ETA: 0s - loss: 0.2023 - accuracy: 0.9230

29/30 [============================>.] - ETA: 0s - loss: 0.2023 - accuracy: 0.9230

30/30 [==============================] - 1s 49ms/step - loss: 0.2020 - accuracy: 0.9232 - val_loss: 0.3036 - val_accuracy: 0.8653


Epoch 19/20


 1/30 [>.............................] - ETA: 0s - loss: 0.1925 - accuracy: 0.9258

 3/30 [==>...........................] - ETA: 0s - loss: 0.1994 - accuracy: 0.9245

 5/30 [====>.........................] - ETA: 0s - loss: 0.1990 - accuracy: 0.9223

 7/30 [======>.......................] - ETA: 0s - loss: 0.1982 - accuracy: 0.9263

 9/30 [========>.....................] - ETA: 0s - loss: 0.1937 - accuracy: 0.9280

11/30 [==========>...................] - ETA: 0s - loss: 0.1905 - accuracy: 0.9281

13/30 [============>.................] - ETA: 0s - loss: 0.1898 - accuracy: 0.9277

15/30 [==============>...............] - ETA: 0s - loss: 0.1882 - accuracy: 0.9284

17/30 [================>.............] - ETA: 0s - loss: 0.1907 - accuracy: 0.9270

19/30 [==================>...........] - ETA: 0s - loss: 0.1907 - accuracy: 0.9277

21/30 [====================>.........] - ETA: 0s - loss: 0.1900 - accuracy: 0.9283

23/30 [======================>.......] - ETA: 0s - loss: 0.1882 - accuracy: 0.9296

25/30 [========================>.....] - ETA: 0s - loss: 0.1877 - accuracy: 0.9304

27/30 [==========================>...] - ETA: 0s - loss: 0.1882 - accuracy: 0.9294

29/30 [============================>.] - ETA: 0s - loss: 0.1885 - accuracy: 0.9296

30/30 [==============================] - 1s 48ms/step - loss: 0.1882 - accuracy: 0.9295 - val_loss: 0.3010 - val_accuracy: 0.8663


Epoch 20/20


 1/30 [>.............................] - ETA: 0s - loss: 0.1963 - accuracy: 0.9277

 3/30 [==>...........................] - ETA: 0s - loss: 0.1844 - accuracy: 0.9362

 5/30 [====>.........................] - ETA: 0s - loss: 0.1831 - accuracy: 0.9332

 7/30 [======>.......................] - ETA: 0s - loss: 0.1831 - accuracy: 0.9350

 9/30 [========>.....................] - ETA: 0s - loss: 0.1815 - accuracy: 0.9355

11/30 [==========>...................] - ETA: 0s - loss: 0.1816 - accuracy: 0.9354

13/30 [============>.................] - ETA: 0s - loss: 0.1763 - accuracy: 0.9392

15/30 [==============>...............] - ETA: 0s - loss: 0.1747 - accuracy: 0.9380

17/30 [================>.............] - ETA: 0s - loss: 0.1739 - accuracy: 0.9374

19/30 [==================>...........] - ETA: 0s - loss: 0.1745 - accuracy: 0.9364

21/30 [====================>.........] - ETA: 0s - loss: 0.1763 - accuracy: 0.9362

23/30 [======================>.......] - ETA: 0s - loss: 0.1751 - accuracy: 0.9372

25/30 [========================>.....] - ETA: 0s - loss: 0.1747 - accuracy: 0.9366

27/30 [==========================>...] - ETA: 0s - loss: 0.1749 - accuracy: 0.9362

29/30 [============================>.] - ETA: 0s - loss: 0.1759 - accuracy: 0.9355

30/30 [==============================] - 1s 49ms/step - loss: 0.1758 - accuracy: 0.9356 - val_loss: 0.3004 - val_accuracy: 0.8657


## Evaluate the model

And let's see how the model performs. Two values will be returned. Loss (a number which represents our error, lower values are better), and accuracy.

In [11]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 1s - loss: 0.3139 - accuracy: 0.8552


loss: 0.314
accuracy: 0.855


This fairly naive approach achieves an accuracy of about 87%. With more advanced approaches, the model should get closer to 95%.

## Further reading

For a more general way to work with string inputs and for a more detailed analysis of the progress of accuracy and loss during training, see the [Text classification with preprocessed text](./text_classification.ipynb) tutorial.